### PR TITLE
feat: add boyer moore algorithm implementation

### PR DESCRIPTION
**Pull request by stoychoX**
#### Checklist
- [x ] Added description of change
- [x ] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x ] Added tests and example, test must pass
- [ ] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [ ] Relevant documentation/comments is changed or added
- [x ] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [ ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x ] I acknowledge that all my contributions will be made under the project's license.

Notes: This implementation is mine, but the algorithm itself is similar to one I saw on geeksforgeeks (as I learned about the algorithm from there). There are other ways to implement the good suffix table, but I truly find this one the best.

### CODE REVIEW

File: strings/boyer_moore.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+ * @file
+ * @brief Boyer-moore's algorithm finding all occurrences of pattern in given
+ * text.
+ * @author [Stoycho Kyosev](https://github.com/stoychoX)
+ */
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <limits>
+#include <vector>
```
Comment (Panquesito7) [inline strings/boyer_moore.cpp:11]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: strings/boyer_moore.cpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+#define APLHABET_SIZE CHAR_MAX
+
+/// @brief A structure representing all the data we need to search the
+/// preprocessed pattern in text.
```
Comment (Panquesito7) [inline strings/boyer_moore.cpp:16]: Please document the functions/anything else similar to this format.
```suggestion
/**
 * @brief A structure representing all the data we need to search the
 * preprocessed pattern in the text.
```

File: strings/boyer_moore.cpp [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    std::vector<size_t> indexes = search(text, pat);
+
+    assert(indexes.size() == 6);
+
+    for (const auto& idx : indexes) {
+        assert(is_prefix(text + idx, "pat", 3));
+    }
+}
+
+void tests() {
```
Comment (Panquesito7) [inline strings/boyer_moore.cpp:153]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
static void tests() {
```

File: strings/boyer_moore.cpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                there was nothing about pat the cloudy sky outside to pat suggest that\
+                        strange and \
+                    mysterious things would pat soon be happening all pat over the \
+                        country.";
+
+    and_test(text);
+    pat_test(text);
+}
+
+int main() { tests(); }
```
Comment (Panquesito7) [inline strings/boyer_moore.cpp:166]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
    tests();
    return 0;
}
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: strings/boyer_moore.cpp [hunk 5]
```diff
@@ -0,0 +1,222 @@
+/**
+ * @file
+ * @brief Boyer-moore's algorithm finding all occurrences of pattern in given
+ * text.
+ * @author [Stoycho Kyosev](https://github.com/stoychoX)
+ */
+#include <cassert>  /// for assert
+#include <cstring>  /// for strlen
+#include <limits>   /// for CHAR_MAX macro
```
Comment (Panquesito7) [inline strings/boyer_moore.cpp:9]: `climits` should be used for this instead.
```suggestion
#include <climits>   /// for CHAR_MAX macro
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: strings/boyer_moore.cpp [hunk 6]
```diff
@@ -0,0 +1,256 @@
+/**
+ * @file
+ * @brief
+ * The Boyer–Moore algorithm searches for occurrences of pattern P in text T by
+ * performing explicit character comparisons at different alignments. Instead of
+ * a brute-force search of all alignments (of which there are n - m + 1),
+ * Boyer–Moore uses information gained by preprocessing P to skip as many
+ * alignments as possible.
+ *
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:9]: Probably needs an @description here

File: strings/boyer_moore.cpp [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+void and_test(const char* text) {
+    strings::boyer_moore::pattern ands;
+    strings::boyer_moore::init_pattern("and", ands);
+    std::vector<size_t> indexes = strings::boyer_moore::search(text, ands);
+
+    assert(indexes.size() == 2);
+    assert(strings::boyer_moore::is_prefix(text + indexes[0], "and", 3));
+    assert(strings::boyer_moore::is_prefix(text + indexes[1], "and", 3));
+}
+
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:219]: I think there should be a documentation comment here

File: strings/boyer_moore.cpp [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+        if (str[i] != pat[i]) {
+            return false;
+        }
+    }
+
+    return true;
+}
+}  // namespace boyer_moore
+}  // namespace strings
+
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:209]: I think there should be a documentation comment here

File: strings/boyer_moore.cpp [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+                    mysterious things would pat soon be happening all pat over the \
+                        country.";
+
+    and_test(text);
+    pat_test(text);
+}
+
+/**
+ * @brief Main function
+ *
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:250]:
```suggestion
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Tested on Gitpod and the tests seem to be failing. Please fix them.

image

**Comment by stoychoX** [comment]
Hmm yes, there was error with size_t overflowing. I think I tested it many times and it did its job I don't know what got wrong this time, but I fixed the type problem.

image

**Comment by stoychoX** [comment]
Found the problem! As I was fixing clang-tidy warnings I changed a type which should not be changed. 

Now tests are fine and I dont see any clang-tidy warnings.

File: strings/boyer_moore.cpp [hunk 10]
```diff
@@ -0,0 +1,263 @@
+/**
+ * @file
+ * @brief
+ * The Boyer–Moore algorithm searches for occurrences of pattern P in text T by
+ * performing explicit character comparisons at different alignments. Instead of
+ * a brute-force search of all alignments (of which there are n - m + 1),
+ * Boyer–Moore uses information gained by preprocessing P to skip as many
+ * alignments as possible.
+ *
+ * @description
```
Comment (Panquesito7) [inline strings/boyer_moore.cpp:10]:
```suggestion
 * @details
```

**Review by Panquesito7 — DISMISSED** [review]
Thanks! 🚀

File: strings/boyer_moore.cpp [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+        arg[str[i]] = str.length() - i - 1;
+    }
+}
+
+/**
+ * @brief A function that initializes pattern
+ *
+ * @param str Text used for initialization
+ * @param arg Initialized structure
+ * @return void
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:149]: consistency is key so we must add return void to everything that returns void
```suggestion
 * @returns void
```

File: strings/boyer_moore.cpp [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+        }
+    }
+}
+
+/**
+ * @brief A function that preprocess the bad char table
+ *
+ * @param str The string being preprocessed
+ * @param arg The bad char table
+ * @return void
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:134]:
```suggestion
 * @returns void
```

File: strings/boyer_moore.cpp [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+        good_suffix;  ///< good suffix table used for [Good Suffix
+                      ///< heuristic](https://www.geeksforgeeks.org/boyer-moore-algorithm-good-suffix-heuristic/?ref=rp)
+};
+
+/**
+ * @brief A function that preprocess the good suffix thable
+ *
+ * @param str The string being preprocessed
+ * @param arg The good suffix table
+ * @return void
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:85]:
```suggestion
 * @returns void
```

File: strings/boyer_moore.cpp [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+        }
+    }
+
+    return true;
+}
+}  // namespace boyer_moore
+}  // namespace strings
+/**
+ * @brief A test case in which we search for every appearance of the word 'and'
+ * @param text The text in which we search for appearance of the word 'and'
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:214]:
```suggestion
 * @param text The text in which we search for appearance of the word 'and'
 * @returns void
```

File: strings/boyer_moore.cpp [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+    std::vector<size_t> indexes = strings::boyer_moore::search(text, ands);
+
+    assert(indexes.size() == 2);
+    assert(strings::boyer_moore::is_prefix(text + indexes[0], "and", 3));
+    assert(strings::boyer_moore::is_prefix(text + indexes[1], "and", 3));
+}
+
+/**
+ * @brief  A test case in which we search for every appearance of the word 'pat'
+ * @param text The text in which we search for appearance of the word 'pat'
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:228]:
```suggestion
 * @param text The text in which we search for appearance of the word 'pat'
 * @returns void
```

File: strings/boyer_moore.cpp [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+
+    assert(indexes.size() == 6);
+
+    for (const auto& idx : indexes) {
+        assert(strings::boyer_moore::is_prefix(text + idx, "pat", 3));
+    }
+}
+/**
+ * @brief Self-test implementations
+ * @return void
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:243]:
```suggestion
 * @returns void
```

File: strings/boyer_moore.cpp [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+
+    and_test(text);
+    pat_test(text);
+
+    std::cout << "All tests have successfully passed!\n";
+}
+
+/**
+ * @brief Main function
+ * @return 0 on exit
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:262]:
```suggestion
 * @returns 0 on exit
```

File: strings/boyer_moore.cpp [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+    return index_storage;
+}
+
+/**
+ * @brief Check if pat is prefix of str.
+ *
+ * @param str pointer to some part of the input text.
+ * @param pat the searched pattern.
+ * @param len length of the searched pattern
+ * @return true if pat is prefix of str. false otherwise.
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:195]:
```suggestion
 * @returns `true` if pat IS prefix of str.
 * @returns `false` if pat is NOT a prefix of str.
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Thank you so much for your contributions! Just a little bit more :heart:

File: strings/boyer_moore.cpp [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+    arg.pat = str;
+    init_bad_char(str, arg.bad_char);
+    init_good_suffix(str, arg.good_suffix);
+}
+/**
+ * @brief A function that implements Boyer-Moore's algorithm.
+ *
+ * @param str Text we are seatching in.
+ * @param arg pattern structure containing the preprocessed pattern
+ * @return (vector of) indexes of the occurrences of pattern in text
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:161]:
```suggestion
 * @return Vector of indexes of the occurrences of pattern in text
```

File: strings/boyer_moore.cpp [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+ */
+void pat_test(const char* text) {
+    strings::boyer_moore::pattern pat;
+    strings::boyer_moore::init_pattern("pat", pat);
+    std::vector<size_t> indexes = strings::boyer_moore::search(text, pat);
+
+    assert(indexes.size() == 6);
+
+    for (const auto& idx : indexes) {
+        assert(strings::boyer_moore::is_prefix(text + idx, "pat", 3));
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:241]: Could you rename the variable idx to index or i  so it's clearer

File: strings/boyer_moore.cpp [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+    }
+}
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void tests() {
+    const char* text =
+        "When pat Mr. and Mrs. pat Dursley woke up on the dull, gray  \
```
Comment (realstealthninja) [inline strings/boyer_moore.cpp:250]: Nitpick Shouldn't it be Mr. pat and Mrs pat

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Just a little bit more!!! 🙂

File: strings/boyer_moore.cpp (hunk 21 above)
Comment (stoychoX) [reply strings/boyer_moore.cpp:250]: This is text I copy-pasted from random book (Harry Potter I think).

I added some "pat" substrings so I could test the search. The text isn't supposed to make any sense 😄

Comment (realstealthninja) [reply strings/boyer_moore.cpp:250]: Ah no wonder! thank you for the clarification!

**Comment by Panquesito7** [comment]
> also @Panquesito7 there seems to be both @ return and @ returns in the contribution guidelines and both are used interchangeably? https://wx-dev.narkive.com/w3HIen0a/doxygen-docs-return-or-returns should we standardize and use either return or returns or keep it as is?

Good question. I think we might want to standardize the use of `return`, rather than `returns`.
I'll create a discussion for this, so we can see what others think about this. 🙂

**Review by tjgurwara99 — APPROVED** [review]
LGTM

File: strings/boyer_moore.cpp [hunk 22]
```diff
@@ -0,0 +1,270 @@
+/**
+ * @file
+ * @brief
+ * The Boyer–Moore algorithm searches for occurrences of pattern P in text T by
```
Comment (Panquesito7) [inline strings/boyer_moore.cpp:4]:
```suggestion
 * The [Boyer–Moore](https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_string-search_algorithm) algorithm searches for occurrences of pattern P in text T by
```

**Review by Panquesito7 — APPROVED** [review]
Amazing work! Thank you. 🚀

**Review by realstealthninja — APPROVED** [review]
Great work! :heart:

**Pull request merged**